### DISCUSSION TITLE
GPT OSS 20b Perf Analysis

### DISCUSSION THREAD

**Discussion by mvasiljevicTT** [tenstorrent/tt-mlir#7334] [Ideas]
Run full model from tt-xla:
```
pytest -x -v tests/benchmark/test_llms.py::test_gpt_oss_20b_tp 
```
Run less layers:
```
pytest -x -v tests/benchmark/test_llms.py::test_gpt_oss_20b_tp --num-layers 4
```
Full model perf report based on [nightly run from 3.3.2026](https://github.com/tenstorrent/tt-xla/actions/runs/22601850826):
<img width="1800" height="2400" alt="perf_bar_gptoss" src="https://github.com/user-attachments/assets/8d7752bf-347a-4baf-81b6-c11b2c981e49" />

Layers in gpt oss alternate between using sliding window and full attention. Perf report for two consecutive layers (one odd, one even):
<img width="1800" height="2400" alt="block_07_double_layer_13_14_summary" src="https://github.com/user-attachments/assets/977ac133-3056-4a2e-be73-6943e9166029" />

Expecting notable improvments:
- [SDPA fusing](https://github.com/tenstorrent/tt-mlir/issues/7086) - expected to highly improve matmuls, remove some of the concat and slice ops
- [RMS fusing](https://github.com/tenstorrent/tt-mlir/issues/7041) - removes expensive fillpad before reduce, tilize/untilize, reshapes...

cc: @odjuricicTT

**Comment by mvasiljevicTT** [comment]
There are 19 `FillPadDeviceOperations` per layer:
- 2: 1 per RMS norm sum - ~300us each - will be solved with rms fusing
- 16: 8 per reduce scatter (one 8\*5us, other 8\*46us)
- 1: for final sum along experts of `routing_weights * expert_output`, ~600us

these are analyzed in [this issue](https://github.com/tenstorrent/tt-mlir/issues/7338).

**Reply by mvasiljevicTT** [reply]
@odjuricicTT device perf files for full model from this analysis:

[gpt_oss_device_perf_csv_260303.zip](https://github.com/user-attachments/files/26110104/gpt_oss_device_perf_csv_260303.zip)

**Comment by mvasiljevicTT** [comment]
[Galaxy device perf initial report](https://raw.githack.com/tenstorrent/tt-xla/aa969f4164d24e0b10169c56ce274d5f9d1fc042/github_pages/perf_reports/2026_03_25_12_13_39/tt_perf/index.html) using claude skills from [this branch](https://github.com/tenstorrent/tt-xla/tree/mvasiljevic/add_profile_skills) (even and odd moe layers are same, this was just experimental, will polish the skills further.)

**Comment by mvasiljevicTT** [comment]
Current implementation of gpt oss on galaxy shards weights on both axis. This is not optimal as then we need distributed rms norm and more inter chip communication in general. 

Changing sharding to use dp on batch axis (4) and tp on model (8) shows better device perf per layer but currently hangs for full model for bs64. ~~For bs128 bumps into same error as current version : https://github.com/tenstorrent/tt-xla/issues/3907~~ It shouldn't. Will be fixed.

~~Will investigate bs64 hang a bit further, but we should focus more on the bs128 issue, as that is the version we care about.~~

Comparative analysis of two current versions for bs64 [on the link](https://raw.githack.com/tenstorrent/tt-xla/9d62794b9944ea1d71d882eb8d6baf7b179343b5/github_pages/perf_reports/compare_reports.html?s=index) from the [mvasiljevic/add_profile_skills](https://github.com/tenstorrent/tt-xla/tree/mvasiljevic/add_profile_skills) branch. 

### Perf Comparator
*also useful as manual profile analysis guidance*

Skills for making those comparisons are on the same branch (still experimental). When the agent is asked to profile the test it will:
- run test with `tracy` on minimal number of layers
- from `ops_perf_results` slice single decode using signposts, and further slice parts that should be analyzed as units (for llama that is one layer, for gpt oss I currently separately analyze odd and even attention sublayers and moe sublayer.) This is the most tricky part as it should cover future graph changes as fusions. And all of this is assuming we are analyzing consecutive device ops (as we should anyway imo, so that is not that of a problem). 
- run `tt-perf-report` on slices, save Performance Report and Stacked Device Time chart
- convert Performance Report to html page, color ops by category, highlight high device time ops in three categories - 50-100us, 100-200us, 200us+ . I made up this categorization, seems useful for now, might change later. Add chart at the bottom of the page
- make overview page that shows some general info about that run - run setup (bs, sharding, layers run), successful fusions, calculated and estimated device and e2e times and t/s/u. Might add some ccl ops analysis here too, or other things that we find useful.

Comparator snapshots for future reference: 

<img width="3408" height="1966" alt="image" src="https://github.com/user-attachments/assets/0bcac28f-e17b-47f2-a94c-f2a8d8822739" />

<img width="3408" height="2052" alt="image" src="https://github.com/user-attachments/assets/29b83fdf-ed5e-4968-8335-8aaed79e6c51" />

fyi @odjuricicTT

**Reply by mvasiljevicTT** [reply]
crossing this with explorer would be super cool. Adding more information to perf report, like what binary op is actually used will also be useful.

**Reply by mvasiljevicTT** [reply]
tried crossing device op slices with ttnn ir to make ir slices without much success. Making slices on ir level first might be better. The goal is to have explorer graph for the slice. Ideas appreciated :D

**Reply by odjuricicTT** [reply]
This is awesome!

It's possible to have something like this as part of explorer. We should precisely define what info from this we would need and then talk to the folks working on tt-explorer.

**Comment by mvasiljevicTT** [comment]
### Gpt oss 20b 4lyr hybrid sharding irs
[ttir_gpt_oss_20b_tp_galaxy_batch_size_64_4lyr_bs64_isl128_run27b3_input_shard_g2_1774613685585.mlir.txt](https://github.com/user-attachments/files/26350867/ttir_gpt_oss_20b_tp_galaxy_batch_size_64_4lyr_bs64_isl128_run27b3_input_shard_g2_1774613685585.mlir.txt)


[ttnn_gpt_oss_20b_tp_galaxy_batch_size_64_4lyr_bs64_isl128_run27b3_input_shard_g2_1774613686389.mlir.txt](https://github.com/user-attachments/files/26350794/ttnn_gpt_oss_20b_tp_galaxy_batch_size_64_4lyr_bs64_isl128_run27b3_input_shard_g2_1774613686389.mlir.txt)

[device_perf_2026_03_27_12_21_16.zip](https://github.com/user-attachments/files/26350925/device_perf_2026_03_27_12_21_16.zip)

**Reply by mvasiljevicTT** [reply]
In this run all weights are sharded like they were in the tp1x8 - attention weights on last dim, and moe on first (expert) dim. 

When changing sharding in moe part to last dim, some already questionable reshapes and fillpad ops are now ridiculously slow, leading to whole moe part now being ~4x slower:

<img width="800"  alt="image" src="https://github.com/user-attachments/assets/c926319b-b043-48af-bfa4-89bf57632c75" />

Total perf with this change is even smaller than the one on the main. So although it doesn't hang this way and moe will be optimized later anyway, this is not acceptable solution.

Anyway, this experiment gives perspective on why we analyze device perf instead of the graph alone, as this fillpad is hidden inside the sum op which wouldn't look that scary on it's own.

**Reply by mvasiljevicTT** [reply]
managed to run full model with correct sharding without hang, e2e is 2.14, which is lower than current main (2.4) even though device perf is better. Trace is not supported for gpt oss yet, that might help.

**Reply by mvasiljevicTT** [reply]
on same machine current main runs at 1.63 sps, so this actually is an improvement. On CI it is 3.24 https://github.com/tenstorrent/tt-xla/actions/runs/23906063735/job/69718665249

Currently, inputs are sharded only for decode, as sharding them for prefill too produces zeros for first 15 tokens. After that, model seems to work fine. Investigating this further.

**Reply by mvasiljevicTT** [reply]
### Sharding issues

When sharding inputs and weights like in the commit in the CI run above, **routing weights on devices 8-31 are all 0**. Since currently all users have same inputs, with current sharding devices with same mod 8 should have same tensors throughout the run. (working branch is mvasiljevic/gpt_oss_dp_tp_sharding but I don't have all the tweaks and logs there)

All ops outside moe router are independent on that axis, so they should not make this kind of error.

Tried dumping inputs and outputs in the router and comparing them with golden to check if it is an op problem, but they all work fine, so the problem is in the graph itself.

_My understanding of how scatter in router currently work matched with actual dumps:_
- For isl 17 and bs 16 per device (64 total), outputs of topk are `[16*17,4]` indices and `[16*17,4]` values.
- indices are concatenated with arange(16*17) and flattened into `[0a, 0b, 0c, 0d, 1a, 1b, ..., 272d],` where `[a, b, c, d]` is every row of indices matrix (all have same inputs so they all have same topK)
- routing weights have nonzero value only for topk experts for each token, and here we need a tensor of those nonzero values and positions. Based on the graph, and the dumps, positions are calculated as `32*t + e`, where `(t,e)` is like `0a` from flattened indices. Then those positions are used on `[32*64*17]` tensor, which is 4x maximum of positions range, so this can't be correct. 
- after scatter, weights are reshaped to `[64*17, 32]`, partitioned to `[16*17, 32]` and then to `[16*17, 4]` preparing it to match outputs for that devices 1/8 of experts.

Couldn't follow exacts parts throughout the stack that form this error. Tried changing position calculation to use range(64*17) instead (as I right this I see this is not reasonable as that x4 in flattened indices comes from top 4 experts, and 64/16 in positions mismatch comes from inputs being sharded to 4 devices which is accidentally the same number). With that change I get pcc 0.92 for some users and 0.6 on others for 6 layers (previously it couldn't even be calculated as all values were 0 and pcc calculation gets division by 0 error), but on full model it diverges.

Tried making repro smaller than whole model with 1 layer, but couldn't completely reproduce the setup to get same kind of error (probably just the skill issue tbh, didn't stay too much on that).

_This might all become irrelevant once MoE part is changed._

Tried injecting `enable_sparse_mlp`, but it breaks on some tensor needing RM instead of tile. Didn't try forcing that, as it doesn't seem that sparse mlp is ready to be used that way.

Using `odjuricic/gptoss-tp-dp` branch seems to work fine even without debug only mode. Each iteration does take 2s+, but we don't get these zero outputs. Didn't deeply analyze the differences yet, on the first glance I see only output sharding which didn't help my case. 

It is possible that that branch already has all we need and just has some overhead that should be fixed.

cc @odjuricicTT

**Comment by odjuricicTT** [comment]
Here are some of the things that should be invesitaged / improved here:

1. SDPA fusing should include the attention mask

2. Should Rope be fused here? Yes if it exists in metal implementation

3. QKV matmuls should be one matmul

4. Why is layernorm kernel so slow?

5. All reduce after O matmul is slow. Why does it contain the FillPad, we should remove it.

6. Bias for O matmul should be fused.

7. Why is residual add so slow, and why is there a reshape before it? Probably some dims need adjusting.

8. Why is there a permute after SDPA? Is that nlp concat heads, if so, why is it not fused?

9. Why is there a permute before SDPA? With the QKV matmuls fused into one there should also be a nlp create heads?

**Comment by odjuricicTT** [comment]
## Sparse MoE Perf Report Analysis

### Raw perf CSVs
- [sparse_moe.csv](https://github.com/tenstorrent/tt-xla/releases/download/untagged-2b905b868b5f941bce02/sparse_moe.csv) — Full model (all layers)
- [sparse_moe_1_1layer.csv](https://github.com/tenstorrent/tt-xla/releases/download/untagged-2b905b868b5f941bce02/sparse_moe_1_1layer.csv) — 1 layer only

### Summary charts

**Only MOE — Stacked Device Time (Total: 8937.9 µs)**
![sparse_moe_summary](https://github.com/tenstorrent/tt-xla/releases/download/untagged-2b905b868b5f941bce02/sparse_moe_summary.png)

**1 decoder layer — Stacked Device Time (Total: 10379.0 µs)**
![sparse_moe_1_1layer_summary](https://github.com/tenstorrent/tt-xla/releases/download/untagged-2b905b868b5f941bce02/sparse_moe_1_1layer_summary.png)

**Comment by mvasiljevicTT** [comment]
### GPT OSS 120b

same as 20b, only with 24->36 layers and 32->128 experts. Except matmuls in MoE part that is still changing, everything should be the same.



[gpt_oss_120b_bs128_sliced_device_perf.zip](https://github.com/user-attachments/files/26540253/gpt_oss_120b_bs128_sliced_device_perf.zip)

[host.tracy.txt](https://github.com/user-attachments/files/26540258/host.tracy.txt)
[ttir_gpt_oss_120b_tp_galaxy_batch_size_128_bs128_isl128_run28ef_g0_1775563108972.mlir.txt](https://github.com/user-attachments/files/26540264/ttir_gpt_oss_120b_tp_galaxy_batch_size_128_bs128_isl128_run28ef_g0_1775563108972.mlir.txt)
[ttnn_gpt_oss_120b_tp_galaxy_batch_size_128_bs128_isl128_run28ef_g0_1775563117268.mlir.txt](https://github.com/user-attachments/files/26540267/ttnn_gpt_oss_120b_tp_galaxy_batch_size_128_bs128_isl128_run28ef_g0_1775563117268.mlir.txt)

perf analysis and reports on [link](https://raw.githack.com/tenstorrent/tt-xla/d2749b96b19ae4609037b65f28d303161b3f37f7/github_pages/perf_reports/2026_04_07_07_11_39/tt_perf/index.html)

**Comment by mvasiljevicTT** [comment]
With sharding from [this PR](https://github.com/tenstorrent/tt-xla/pull/4205), MoE part gets extremely slow, mainly due to point-to-point blocks that switch between dp4(/tp8) attention and tp4(x8) moe, as seen on device perf for whole layer:
<img width="600"  alt="odd_full_summary" src="https://github.com/user-attachments/assets/84c528be-4316-4c0f-b99e-6dde279f7ff8" />

Attention now takes only 6% device time of the whole layer:
<img width="600"  alt="odd_attention_summary" src="https://github.com/user-attachments/assets/6e909467-6aa5-4d57-bdbe-ec7db33f84a7" />

detailed analysis on the [linik](https://raw.githack.com/tenstorrent/tt-xla/mvasiljevic/gpt-oss-dp-tp-working-sharding-pages/github_pages/perf_reports/2026_04_14_13_25_34/tt_perf/index.html).

Profiled for gpt oss 20b, bs64 for clear comparison with previous versions.

Even with MoE being that slow, based on device perf alone it is estimated to generate 4.6 t/s/u, and e2e it runs 0.6 t/s/u, which is suggesting major e2e overhead that is not present in plain dual axis weight shard previously used.

**Reply by mvasiljevicTT** [reply]
Huge FillPad op in attention comes from reduce scatter after last matmul, which might be resolved on its own with new rs implementation that is not yet uplifted on tt-xla

**Reply by mvasiljevicTT** [reply]
First user have somewhat ok output:
```
User 0: "

But not.

But there is "Here is an exhaustive list of the best practices for writing clean code:".

But not.

But there is "Here is an exhaustive list of the best practices for writing clean code:".

But not.

But there is "Here is an exhaustive list of the best practices for writing clean code:".

But not.

But there is "Here is an exhaustive list of the best practices for writing clean code:".

But not.

But there is "Here is an exhaustive list of the best practices for writing clean code:".
```
and others are mostly random.


Looking at scatter op that was previously analyzed, this implementation seems to have similar problem with indices calculated as 
```
flat_index = row * 32 + col
row <-- repeat 4 arange(0,272) 
```
having flat index in range `[0, 32*272)`, used to index tensor with `32*4*272` elements. This is indicating misalignment in experts considered as topk.

**Reply by mvasiljevicTT** [reply]
## GPT OSS 120b bs128

Interestingly, attention for 120b bs128 (same as attention for 20b bs128) is faster as it has all same ops but without slow fillpad:

<img width="4572" height="1096" alt="image" src="https://github.com/user-attachments/assets/842eb58b-611d-47af-8b9d-29bb1d08e444" />

while MoE block is ~4x slower, which is expected as it has 4x experts (and also bs128 vs bs64):
<img width="600"  alt="image" src="https://github.com/user-attachments/assets/a062d1e7-00ef-412a-8604-5c4ca6df0c36" />

making attention 1.2% of the whole layer 🤯 

[detailed analysis and comparisons](https://raw.githack.com/tenstorrent/tt-xla/e65d9729e7ac1754b10f55ac640a198d474ffd2b/github_pages/perf_reports/2026_04_14_16_30_31/tt_perf/index.html)

Just the fillpads, slices and point-to-point ops, take 2/3 of the time, so there is a lot of room for improvement for new MoE implementation.

**Reply by mvasiljevicTT** [reply]
with current implementation pcc is ok only for first 32 users (first 1/4 of the batch) and output is ok only for user 0. 

Changing `python_package/tt_torch/torch_overrides.py::_router_forward` to use `values*one_hot(indices)` instead of `zeros.scatter_(indices,values)` to bypass scatter op, pcc (on prefill) becomes 0.99 for all the users even though the outputs are mostly garbage except for the users 0, 32, 64, and 96 (first on each device).

**Reply by mvasiljevicTT** [reply]
Issue of only first user of every device having correct values was because mask in SDPA was not broadcasted, so only first user used correct one.

Fix on [branch](https://github.com/tenstorrent/tt-mlir/tree/mvasiljevic/sdpa-broadcast-mask)

This was a wider issue that affected some other multichip models too.

**Reply by mvasiljevicTT** [reply]
Still sure how single chip models, for example llama, work ok even with [1,1,H,S] shape. Might be because it has 32 heads, instead of 4/8 in multichip, so the mask is already repeated enough, or they just have less error throughout the model that would amplify it. 

It is also possible that model works ok with 0 mask if KV cache is correct aka 0 on future places. If that is the case, it is likely that multichip models have more noise on empty places for whatever reasons, so with either mask or KV cache being nonzero it gets incorrect fast. That would also explain why it worked better with injected KV cache, although without testing I am not sure if it was zeros in the cache, or some random difference in the mask memory position that contributed to that.

In some tests I could see that user 2 was almost ok and as it becomes more gibberish. This was because first users are closest to having all zero mask? 

It would be interesting to investigate this more with python tests on cpu.

**Reply by mvasiljevicTT** [reply]
python tests confirm - zero mask and correct KV cache changes logits, but in a scaled way, so argmax will still always choose  same tokens. This is why single chip tests didn't seem to be affected by this.

With dummy weights in the python test, top5 metric drops with the number of layers, up to 40% for layer 32. 

Even though all outputs are same, checking pcc for all users might show some difference in 1st vs other users
